odoo/addons/base/models/res_users.py

```diff
@@ -1427,3 +1423,116 @@ def change_password_button(self):
+        type(self).SELF_READABLE_FIELDS = self.SELF_READABLE_FIELDS + [
+            'api_keys_only_explicit', 'api_keys_only', 'api_key_ids'
+        ]
+        return init_res
+
+    def write(self, vals):
+        r = super().write(vals)
+        # needs to invalidate the UID cache as this potentially makes
+        # previously valid auth invalid
+        if 'api_keys_only_explicit' in vals:
+            self.clear_caches()
+        return r
+
+    @api.depends('api_keys_only_explicit')
+    def _compute_keys_only(self):
+        for u in self:
+            u.api_keys_only = u.api_keys_only_explicit
+
+    def _check_credentials(self, password, user_agent_env):
+        if user_agent_env['interactive']:
+            return super()._check_credentials(password, user_agent_env)
+
+        if not self.env.user.api_keys_only:
+            try:
+                return super()._check_credentials(password, user_agent_env)
+            except AccessDenied:
+                pass
+
+        self.env.cr.execute('''
+        SELECT key FROM res_users_apikeys WHERE user_id = %s AND index = %s
+        ''', [self.env.uid, password[:INDEX_SIZE]])
+        for [key] in self.env.cr.fetchall():
+            if KEY_CRYPT_CONTEXT.verify(password, key):
+                return
+
+        raise AccessDenied()
+
+class APIKeys(models.Model):
+    _name = _description = 'res.users.apikeys'
+    _auto = False # so we can have a secret column
+
+    name = fields.Char("Description", required=True, readonly=True)
+    user_id = fields.Many2one('res.users', index=True, required=True, readonly=True, ondelete="cascade")
+
+    def init(self):
+        self.env.cr.execute("""
+        CREATE TABLE IF NOT EXISTS {table} (
+            id serial primary key,
+            name varchar not null,
+            user_id integer not null REFERENCES res_users(id),
+            index varchar({index_size}) not null CHECK (char_length(index) = {index_size}),
+            key varchar not null
+        );
+        CREATE INDEX ON {table} (user_id, index);
+        """.format(table=self._table, index_size=INDEX_SIZE))
+
+    def unlink(self):
+        self.clear_caches()
+        return super().unlink()
+
+class APIKeyDescription(models.TransientModel):
+    _name = _description = 'res.users.apikeys.description'
+
+    name = fields.Char("Description", required=True)
+
+    def make_key(self):
+        # only create keys for users who can delete their keys
+        self.env['res.users.apikeys'].check_access_rights('unlink')
+        # no need to clear the LRU when *adding* a key, only when removing
+        k = binascii.hexlify(os.urandom(API_KEY_SIZE)).decode()
+        self.env.cr.execute("""
+        INSERT INTO res_users_apikeys (name, user_id, key, index)
+        VALUES (%s, %s, %s, %s)
+        RETURNING id
+        """, [self.sudo().name, self.env.user.id, hash_api_key(k), k[:INDEX_SIZE]])
```

Comment:
@odony I'm hashing the hexlified version of the key here rather than the actual binary data, should not make any real difference and avoids having to unhexlify it on verification but it's kinda weird, what's your take?